nwbwidgets/utils/timeseries.py

```diff
@@ -137,29 +137,97 @@ def timeseries_time_to_ind(node: TimeSeries, time, ind_min=None, ind_max=None) -
         return int(np.ceil((time - starting_time) * node.rate))
 
 
-def align_by_times(timeseries: TimeSeries, starts, stops):
+def bisect_array_by_times(data: np.array, timestamps: np.array, starts, duration:float, traces=None):
     """
-    Args:
+    Parameters
+    ----------
+    data: np.array
+        array to bisect
+    timestamps: np.array
+        timestamps corresponding to dim 0 of data
+    starts: iterable
+        time at which to bisect
+    duration: float
+        duration of window after start
+    traces: int
+        index into the second dim of data
+    Returns
+    -------
+    out: list
+        list with bisected arrays from data
+    """
+    out = []
+    for start in starts:
+        idx_start =bisect(timestamps, start)
+        idx_stop = bisect(timestamps, start + duration, lo=idx_start)
+        if len(data.shape) > 1 and traces is not None:
+            out.append(data[idx_start:idx_stop, traces])
+        else:
+            out.append(data[idx_start:idx_stop])
+    return out
+
+
+def align_by_times_with_timestamps(timeseries: TimeSeries, starts, duration: float, traces=None):
+    """
+    Parameters
+    ----------
+    timeseries: TimeSeries
+        timeseries with variable timestamps
+    starts: array-like
+        starts in seconds
+    duration: float
+        duration in seconds
+    Returns
+    -------
+    out: list
+        list: length=(n_trials); list[0]: array, shape=(n_time, ...)
+    """
+    assert timeseries.timestamps is not None, 'supply timeseries with timestamps'
+    return bisect_array_by_times(timeseries.data, timeseries.timestamps, starts, duration, traces)
+
+
+def align_by_times_with_rate(timeseries: TimeSeries, starts, duration: float, traces=None):
+    """
+    Parameters
+    ----------
+    timeseries: TimeSeries
+        timeseries with variable timestamps
+    starts: array-like
+        starts in seconds
+    duration: float
+        duration in seconds
+    Returns
+    -------
+    out: list
+        list: length=(n_trials); list[0]: array, shape=(n_time, ...)
+    """
+    assert timeseries.rate is not None, 'supply timeseries with start_time and rate'
+    ts = np.arange(timeseries.data.shape[0])*1/timeseries.rate + timeseries.starting_time
+    return bisect_array_by_times(timeseries.data, ts, starts, duration, traces)
```

Comment:
If you are given a starting time and rate, you should calculate the indices of `data` directly, not create a timeseries and bisect it.